treeherder/model/management/commands/cycle_data.py

```diff
@@ -7,17 +7,98 @@
                                      JobType,
                                      Machine,
                                      Repository)
-from treeherder.perf.models import PerformanceDatum
+from treeherder.perf.models import (PerformanceDatum,
+                                    PerformanceDatumManager)
+
+TREEHERDER = 'treeherder'
+PERFHERDER = 'perfherder'
+DATA_PRODUCERS = [TREEHERDER, PERFHERDER]
+
+
+class DataCycler:
+    source = ''
+
+    def __init__(self, days, chunk_size, sleep_time, is_debug=None,
+                 logger=None, **kwargs):
+        self.cycle_interval = datetime.timedelta(days=days)
+        self.chunk_size = chunk_size
+        self.sleep_time = sleep_time
+        self.is_debug = is_debug or False
+        self.logger = logger
+
+    def cycle(self, repository):
+        pass
+
+
+class TreeherderCycler(DataCycler):
+    source = TREEHERDER.title()
+
+    def cycle(self):
+        for repository in Repository.objects.all():
+            self.logger.debug("Cycling repository: {0}".format(repository.name))
+            rs_deleted = Job.objects.cycle_data(repository,
+                                                self.cycle_interval,
+                                                self.chunk_size,
+                                                self.sleep_time)
+            self.logger.debug("Deleted {} jobs from {}".format(rs_deleted,
+                                                               repository.name))
+
+        self.remove_leftovers()
+
+    def remove_leftovers(self):
+        used_job_type_ids = Job.objects.values('job_type_id').distinct()
+        JobType.objects.exclude(id__in=used_job_type_ids).delete()
+
+        used_job_group_ids = Job.objects.values('job_group_id').distinct()
+        JobGroup.objects.exclude(id__in=used_job_group_ids).delete()
+
+        used_machine_ids = Job.objects.values('machine_id').distinct()
```

Comment:
Can you return multiple values the first time instead of querying `Job` three times?

Reply:
> Can you return multiple values the first time instead of querying `Job` three times?

I think that querying `Job` 3 times is going to make many of the deletions that follow redundant. For example, we could have multiple (`job_type_id`, `job_group_id`, `machine_id`) tuples with job_group_id that repeat very often. We would delete them multiple times.